File: advent_of_code_2023/day1/day1.ts
import {readFileSync} from "fs";

function convertWords(line: string) {
    return line.replace(/one/g, "one1one")
    .replace(/two/g, "two2two")
    .replace(/three/g, "three3three")
    .replace(/four/g, "four4four")
    .replace(/five/g, "five5five")
    .replace(/six/g, "six6six")
    .replace(/seven/g, "seven7seven")
    .replace(/eight/g, "eight8eight")
    .replace(/nine/g, "nine9nine");
}

function getDigits(line: string, part2=false): number {
    if (part2) {
        line = convertWords(line);
    }
    let digits = [-1,-1];
    for (let c of line) {
        if (c.match("[0-9]")) {
            if (digits[0] === -1) {
                digits[0] = parseInt(c);
            }
            digits[1] = parseInt(c);
        }
    }
    if (digits[0] === -1) {
        return 0;
    }
    return 10 * digits[0] + digits[1];
}

function solve(lines: string[]): void {
    let sum1 = 0;
    let sum2 = 0;
    for (let line of lines) {
        sum1 += getDigits(line);
        sum2 += getDigits(line, true);
    }
    console.log("Part 1:", sum1);
    console.log("Part 2:", sum2);
}

function main() {
    let lines: string[] = readFileSync("input1.txt", "utf-8").split('\n').map(x => x.trim());
    console.time("Time taken");
    solve(lines);
    console.timeEnd("Time taken");
}

main();

File: advent_of_code_2023/day1/day1.py
def arrayise(filename):
    array = []
    f = open(filename, "r")
    for line in f:
        array.append(line.strip())
    return array

def solve(array):
    sum = 0
    for line in array:
        first_digit = -1
        last_digit = 0

        while len(line) > 0:
            c = line[0]
            if c.isdigit():
                if first_digit == -1:
                    first_digit = int(c)
                last_digit = int(c)
            elif line.startswith("one"):
                if first_digit == -1:
                    first_digit = 1
                last_digit = 1
            elif line.startswith("two"):
                if first_digit == -1:
                    first_digit = 2
                last_digit = 2
            elif line.startswith("three"):
                if first_digit == -1:
                    first_digit = 3
                last_digit = 3
            elif line.startswith("four"):
                if first_digit == -1:
                    first_digit = 4
                last_digit = 4
            elif line.startswith("five"):
                if first_digit == -1:
                    first_digit = 5
                last_digit = 5
            elif line.startswith("six"):
                if first_digit == -1:
                    first_digit = 6
                last_digit = 6
            elif line.startswith("seven"):
                if first_digit == -1:
                    first_digit = 7
                last_digit = 7
            elif line.startswith("eight"):
                if first_digit == -1:
                    first_digit = 8
                last_digit = 8
            elif line.startswith("nine"):
                if first_digit == -1:
                    first_digit = 9
                last_digit = 9
            line = line[1:]
        sum += first_digit * 10 + last_digit
    print(sum)


if __name__ == '__main__':
    filename = "input1.txt"
    arr = arrayise(filename)
    solve(arr)